src/core/markdownASTToString.ts

```diff
@@ -82,35 +82,48 @@ function markdownContentASTToString(nodes: SemanticMarkdownAST[], options?: Conv
                 case 'italic':
                 case 'strikethrough':
                 case 'link':
-                    const isLastWhitespace = /\s/.test(markdownString.slice(-1));
-                    const textNodeTypes = ['text', 'bold', 'italic', 'strikethrough', 'link'];
-                    const isStartsWithWhiteSpace = textNodeTypes.includes(node.type) && /\s/.test(node.content.slice(0, 1)[0] as string);
-
-                    if (!isLastWhitespace && node.content !== '.' && !isStartsWithWhiteSpace) {
-                        markdownString += ' ';
-                    }
+                    const prevNode = index > 0 ? nodes[index - 1] : null;
+                    const nextNode = index < nodes.length - 1 ? nodes[index + 1] : null;
+                    const isPrevNodeFormatting = prevNode && ['bold', 'italic', 'strikethrough', 'link'].includes(prevNode.type);
+                    const isNextNodePunctuation = nextNode && nextNode.type === 'text' && /^[.,!?;:]/.test(nextNode.content);
 
                     if (node.type === 'text') {
-
+                        if (!isPrevNodeFormatting && !isNextNodePunctuation && !/^\s/.test(node.content) && markdownString.length > 0 && !/\s$/.test(markdownString)) {
```

Comment:
/\s$/.test(markdownString) is an O(n) operation and could be quite expensive with large documents.. please use /\s/.test(markdownString.slice(-1) instead.
Also, please move it a bit up and give this a meaningful name